hoomd/data/local_access.py

```diff
@@ -0,0 +1,453 @@
+        object.__setattr__(self, '_accessed_fields', dict())
+
+
+class _ParticleLocalAccess(_LocalAccess):
+    @property
+    @abstractmethod
+    def _cpp_cls(self):
+        pass
+
+    _fields = {
+        'position': 'getPosition',
+        'typeid': 'getTypes',
+        'velocity': 'getVelocities',
+        'mass': 'getMasses',
+        'acceleration': 'getAcceleration',
+        'orientation': 'getOrientation',
+        'angular_momentum': 'getAngularMomentum',
+        'moment_of_inertia': 'getMomentsOfInertia',
+        'charge': 'getCharge',
+        'diameter': 'getDiameter',
+        'image': 'getImages',
+        'tag': 'getTags',
+        'rtag': 'getRTags',
+        'rigid_body_id': 'getBodies',
+        'net_force': 'getNetForce',
+        'net_torque': 'getNetTorque',
+        'net_virial': 'getNetVirial'}
+
+    def __init__(self, state):
+        super().__init__()
+        object.__setattr__(
+            self, '_cpp_obj',
+            self._cpp_cls(state._cpp_sys_def.getParticleData()))
+
+
+class ParticleLocalAccessCPU(_ParticleLocalAccess):
+    """Class for directly accessing HOOMD-blue particle data on CPU.
+
+    Attributes:
+        type ((N_particles) `hoomd.array` object of ``float``): The integer
+            type of a particle
+        tag ((N_particles) `hoomd.array` object of ``int``): the tag of a
+            particle.  HOOMD-blue uses spacial sorting to improve cache
+            efficiency in particle look-ups. This means the ordering of the
+            array changes.  However, particle tags remain constant. This
+            means that if ``particles.tag[0]`` is 1, then later whatever
+            index has a value of 1 later in the simulation is the same
+            particle.
+        rtag ((N_particles_global) `hoomd.array` object of ``int``): the
+            reverse tag of a particle. This means that the value
+            ``particles.rtag[0]`` represents the current index for particle
+            with tag 0.
+        position ((N_particles, 3) `hoomd.array` object of ``float``):
+            particle positions
+        image ((N_particles, 3) `hoomd.array` object of ``int``): the
+            periodic image a particle occupies
+        velocity ((N_particles, 3) `hoomd.array` object of ``float``):
+            particle velocities (note that changing velocities may or may
+            not effect the future trajectory)
+        acceleration ((N_particles, 3) `hoomd.array` object of ``float``):
+            particle accelerations (note that changing accelerations may or
+            may not effect the future trajectory)
+        mass ((N_particles) `hoomd.array` object of ``float``): particles'
+            masses
+        'orientation ((N_particles, 4) `hoomd.array` object of ``float``):
+            particle orientations expressed as quaternions
+        angular_momentum ((N_particles, 4) `hoomd.array` object of ``float``):
+            particle angular momenta expressed as quaternions
```

Comment:
Well we store it as a quaternion
```
std::vector< quat<Real> > angmom;          //!< angular momentum quaternion
...
GlobalArray< Scalar4 > m_angmom;               //!< Angular momementum quaternion for each particle
````

Reply:
@joaander I've never actually had a need to access the angular momenta in a 3D simulation, so I guess I've never really thought about this. But exposing angular momenta as a quaternion seems very confusing to me, I had completely forgotten that this was the case. The quaternion representation is due to the integration schemes that Jens implemented for rigid bodies IIRC, because Trotter-Suziki factorization of the rotation Liouvillian is much more stable in the quaternion representation, right? However, I'd consider myself a reasonably competent MD user, and without looking at the papers describing this scheme I don't know how to convert this to a more familiar representation of angular momentum. Is it as simple as converting to an axis-angle representation? If so, how is the magnitude encoded (given that this is a unit quaternion), since the angle is bounded?

This question seems mostly unrelated to this PR and I don't think we should address it here, but I'm wondering if we need to consider this a little more closely for the final 3.0 API. If you think any change (or at least further discussion) is merited, I would suggesting making a new issue from this thread and then resolving this conversation so that it doesn't hold up this PR. However, I'd be concerned about the usefulness of the angular momenta in their current form.

Reply:
These data structures provide access to the internal fields stored by HOOMD. We can't introduce alternate computed forms of these quantities without creating copies. We could improve the documentation and perhaps even include Python code showing users how to convert to a vector representation. Here is the documentation for that: https://hoomd-blue.readthedocs.io/en/stable/aniso.html#quaternions-for-angular-momentum

Reply:
I had forgotten to check that page. OK, it's good that we have that at least. I agree that there's no way to do this without making a copy, but perhaps it would be worth introducing a small helper function that does that for the user if they ask for it? Otherwise yes, including a snippet in the docs could be helpful.